Style the html manual, changing type and layout

charlesetc opened the pull request:
Hi!

So this commit makes major visual changes to the OCaml manual. I'm very happy to take any feedback and make changes.

I've got [a preview up](http://preview_docs.narwhal.space/index.html) of these changes. This is compared to [the current version](https://caml.inria.fr/pub/docs/manual-ocaml/).

The header font is [Fira Sans](https://mozilla.github.io/Fira/). I checked it into this repository, so the font would be served from OCaml's servers. There is an argument to be made for using Google Fonts instead, since it can make things faster through caching. I'd be interested in hearing what people think about that.

The only "hack" here is the following snippet:
```
\newstyle{body >
          div:first-child >
          span:first-child >
          span:first-child}{
    font-family: "Fira Sans", sans-serif;
}
```

Which selects the "The OCaml System Release 4.08 ..." heading to make it a different font. Unfortunately there's no id or class by which to select this — if anyone knows how to add one that'd be awesome!

In reference to the recent change (https://github.com/ocaml/ocaml/commit/49abafc5872f6be347cdb72a8a4ecccd229b6118) of the manual, this keeps the max-width for the body, but adds a `margin: auto` so that the body is centered. I think this looks slightly better and also more like the current manual. On small screens, when max-width doesn't apply, this also adds a percentage padding. You can go to [preview_docs.narwhal.space](http://preview_docs.narwhal.space) on a phone to see how it looks.

Anyways, I'm glad to make any changes people see fit :cactus:

charlesetc committed:
Style the html manual, changing type and layout

Add a new font for the headers, change the link colors,
re-center it, and improve support for mobile viewing.

xavierleroy commented:
I'm glad to see efforts to improve the look and legibility of the online manual.  However something is very wrong with the OCaml code snippets, e.g. the first example of section 1.1, `1+2*3`.  The original rendering has colors and a `#` prompt to distinguish input from output.  The new style has none of that and is unreadable.

charlesetc commented:
You're right! I'll see what I can do about that.

charlesetc commented:
It's fixed now! I had the preview showing outdated code

Octachron commented on manual/manual/manual.tex:
```diff
@@ -14,6 +14,18 @@
+}
+\fi
+
 \input{macros.tex}
```
This  `\input{macros.tex}` include should be before the redefinition of the `\@meta` (since it is the one that defines `\ifouthtml`).

charlesetc replied:
Thanks! It should be fixed now

charlesetc committed:
Fix the order of \input{macros.tex} and meta tag

Drup commented:
I really like it, but as someone with terrible eyesight, I would prefer if the background was not pure white. While it looks good and modern, pure white background is very tiring for the eyes. I also tend to like when code blocks have specific background.  
For example, [RWO](https://realworldocaml.org/v1/en/html/a-guided-tour.html) uses a slightly beige background and darker code blocks, which is far less aggressive.

Similarly, you might want to tone down the electric blue a bit.

Misc remarks:
- Some definition overflow, such as http://preview_docs.narwhal.space/extn.html#sec236
- I believe the links to modules in the documentation are not rendered correctly. See `Int32` in http://preview_docs.narwhal.space/core.html#sec549
- The small yellow arrows certainly have a vintage charm, but do not really correspond to the style anymore. :)

The jump to the library documentation is very jarring now. I suppose that is your next step?

bluddy commented:
That's some good css hacking. Adding classes isn't possible without modifying hevea, which isn't on github. I'm really impressed by what you've been able to accomplish just using this method.

charlesetc committed:
Add backgrounds and move to darker blue links

pmetzger commented:
I think that using Google's free Fonts CDN for Fira Sans will be better than serving it off of our own servers — it will load faster.

pmetzger commented:
> Unfortunately there's no id or class by which to select this — if anyone knows how to add one that'd be awesome!

So I've also been working on doing something similar, and Lambda Soup can be used to postprocess the manual quite nicely for this purpose. One of the changes I've made in my own hacking has been to replace the arrows at the tops and bottoms of the pages to include SVG arrows in their place. You can also nicely alter the CSS classes in the way you wish. Code to do this is quite compact and easy.

pmetzger commented:
Oh, and in my own experiments I've used Fira Sans for the headers just as you did, and EB Garamond for the body text. EB Garamond is one of the nicest serif fonts available on Google Fonts. Try it, it's quite nice.

pmetzger commented:
On the background, I find the beige a little too tinted. Perhaps F7F7F7 instead?

charlesetc committed:
Format stdlib pages, use font cdn, switch bg color

pmetzger commented:
> The small yellow arrows certainly have a vintage charm, but do not really correspond to the style anymore.

So I have two options to suggest there, both easily added by the use of Lambda Soup to edit the HTML.

1. Just say `Back`, `Up`, `Forward`.
2. Use Google's "Material Design" SVG arrow icons, which are very nice and scale perfectly to every resolution since they're vector art:

https://github.com/google/material-design-icons/blob/master/navigation/svg/design/ic_arrow_back_24px.svg

https://github.com/google/material-design-icons/blob/master/navigation/svg/design/ic_arrow_upward_24px.svg

https://github.com/google/material-design-icons/blob/master/navigation/svg/design/ic_arrow_forward_24px.svg

pmetzger commented:
On `style.css` for stdlib, I was working on this yesterday and struggling to get everything quite right because the CSS classes are so eccentric, but it should be possible to get something quite nice and consistent with the rest of the manual by using a touch of Lambda Soup to postprocess and change the class structure.

pmetzger commented:
BTW, @aantron was helping me with Lambda Soup assistance, so I'm tagging him here in case there are Lambda Soup questions.

charlesetc commented:
Haha thanks @bluddy! @Drup thanks for the great suggestions! I've added a light grey background (@pmetzger  F7F7F7, thanks!) to the body, and a beige background + light outline to code blocks. This code block change is another big visual change, so we should be sure about it. I also changed that light blue on visited links to a darker blue which should be better. I'm not sure what's wrong with the links to modules like `Int32` — it looks to my like it's the same in the current version. Am I missing something?

Unfortunately there's already a tiny bit of definition overflow in the current manual. Bumping up the font size, and limiting the body adds to that. It's also arranged with tables so I haven't yet found a way to get them to wrap around or scroll... but I'll keep trying.

Also good point about the standard library formatting! I've just added styling that does it. See [this Str module](http://preview_docs.narwhal.space/libref/Str.html).

I'm actually kind of fond of the arrows, but they are obviously dated. I don't know anything about Lambda Soup but it does seem like it would take some kind of post-processing to do it. @pmetzger everything you're saying about lambda soup sounds awesome and is definitely what this needs!

I just tried EB Garamond and it's definitely a nice font but looks a little strange here so I left it out for now ![eb](https://user-images.githubusercontent.com/5167293/39610609-3706b9f4-4f17-11e8-8f68-a278793ed44c.png)

charlesetc commented:
Oh also I switched to the Google CDN :)

pmetzger commented:
The reason it looked strange in the libref stuff is because of the horrid class problems in the generated ocamldoc. Try using EB Garamond only in the hevea generated manual for now and you'll be happier. We'll need Lambda Soup to clean up the ocamldoc classes in order to do real CSS font styling there.

pmetzger commented:
So Lambda Soup may be found at: https://github.com/aantron/lambda-soup

You will note that the documentation for Lambda Soup is itself postprocessed with a Lambda Soup program that is given in the github repository.

pmetzger commented:
(I'll be back tomorrow. Big kudos on doing this work!)

charlesetc commented:
Wow Lambda soup looks awesome and it's ocamldoc site is beautiful!

Thanks :smile:  I'll also be back tomorrow to:

* Try to fix the overflowing BNF's
* Add EB Garamond (having trouble atm)

gasche commented:
For the record, I'm not completely excited at the idea of post-processing generated documentation. Using Lambda Soup is fine for experiments, but if we decide that we like a certain style, why not change the code generation instead to get it from the scratch? Post-processing would make the manual/documentation build harder to maintain.

Octachron commented:
@charlesetc :
> Which selects the "The OCaml System Release 4.08 ..." heading to make it a different font. Unfortunately there's no id or class by which to select this — if anyone knows how to add one that'd be awesome!

Sorry, I missed this point on first reading. One could fix this problem by making the latex code more semantic and replacing the `\begin{center}...\end{center}` used in the main title by a new `maintitle` environment and then adding the right definition to `macros.tex` and `macros.hva`: [example](https://github.com/Octachron/ocaml/commit/ede52ab936c5061904a79ddb1d7181c1f9cdf841)

Drup commented:
@charlesetc Thanks, this is much nicer!

I agree with @gasche, in particular regarding the API documentation. We can tweak the output of ocamldoc whichever way we want, might as well avoid doing postprocessing. Also, improvements in ocamldoc styling can easily benefit odoc as well. 

Notably, Having a side panel with an index for the manual would be really cool.

pmetzger commented:
> For the record, I'm not completely excited at the idea of post-processing generated documentation. Using Lambda Soup is fine for experiments, but if we decide that we like a certain style, why not change the code generation instead to get it from the scratch?

That requires a lot more learning to do. Fixing the HTML is easy. Fixing ocamldoc requires learning how the inside of ocamldoc works and that's much less obvious. I would recommend we not postpone fixing up the appearance until that work gets done. Instead, I would suggest we do this in two steps:

1. We do some postprocessing now, just to figure out what we want and get things looking nice for the short term. That gets committed for now, because it will make the documentation more readable for people now. The nice thing about code is it isn't stone, we can get rid of it at will in a month or six months.
2. A volunteer who understands ocamldoc and hevea fixes them up to generate better things, and then the postprocessor is thrown out, but they don't have to rush to do this because we have the postprocessor as a crutch for the moment.

pmetzger commented:
> Notably, Having a side panel with an index for the manual would be really cool.

We can do that pretty quickly with Lambda Soup, and we can fix it "correctly" over time. @aantron has an example in his own documentation for Lambda Soup.

Chris00 commented:
> We do some postprocessing now, just to figure out what we want and get things looking nice for the short term.

For the library documentation (as opposed to the manual itself), if there are hesitations to merge post-processing in the repo, an opam package with a small tool would be nice (and useful for other people as well).

gasche commented:
> I would suggest we do this in two steps:
>
> 1. We do some postprocessing now, just to figure out what we want and get things looking nice for the short term. That gets committed for now, because it will make the documentation more readable for people now. The nice thing about code is it isn't stone, we can get rid of it at will in a month or six months.
> 2. A volunteer who understands ocamldoc and hevea fixes them up to generate better things, and then the postprocessor is thrown out, but they don't have to rush to do this because we have the postprocessor as a crutch for the moment.

What will predictably happen in that scenario is that no one will ever do step (2), and we will end up with a build system that is even more baroque than it is today. We will still be there maintaining the manual build ten years from now, and I think it really is in our collective interest for me to look a bit more annoying right now, and try to get someone (not necessarily the PR author) motivated enough to make the rendering fixable at the source *now* and not in six renewable months.

Octachron commented:
Another point is that is not like hevea or ocamldoc are black magic that noone understands. Using documentation post-processing in the explorationphase sounds like a good idea. But once there is a clear target, I would also argue that the changes should be backported to hevea or ocamldoc. For instance, I imagine that it would benefit to all hevea users to have a more modern choice of navigation arrows.

nojb commented:
> What will predictably happen in that scenario is that no one will ever do step (2), and we will end up with a build system that is even more baroque than it is today. We will still be there maintaining the manual build ten years from now, and I think it really is in our collective interest for me to look a bit more annoying right now, and try to get someone (not necessarily the PR author) motivated enough to make the rendering fixable at the source now and not in six renewable months.

Seconded. The manual is hard enough to build under Windows, no need to make it impossible to do so.

If someone could explain what are the problems/shortcomings with the current `ocamldoc`/`hevea` output, people who are familiar with them could look into what is required to do the necessary modifications.

gasche commented:
@charlesetc: could you add a Changes entry, in the 4.07 part, section "Manual and documentation", crediting people that contributed to the patch?

Also, it would be nice to integrate @Octachron's suggestion in https://github.com/ocaml/ocaml/pull/1757#issuecomment-386525751 on how to get a proper id/class for the main title.

(My gut feeling is that the PR could be merged after this. In particular, I'm at peace with the way the extra `<meta>` tag is handled.)

Octachron commented:
> In particular, I'm at peace with the way the extra <meta> tag is handled.

( in fact, this fragment is directly extracted from the [hevea manual](http://hevea.inria.fr/doc/manual024.html)  on how on to add meta tags to the `<head>` tag )

pmetzger commented:
> What will predictably happen in that scenario is that no one will ever do step (2), and we will end up with a build system that is even more baroque than it is today.

My fear is if we don't do it in steps, it won't happen, and we'll end up with the ugly manual forever!

How about we compromise by finding an expert on the ocamldoc code who is willing to do the work with @charlesetc (and possibly me) so that we get the best of both?

pmetzger commented:
BTW, if we want to change the arrows to SVG in `hevea` itself, how do we do that?

pmetzger commented:
@nojb:
> If someone could explain what are the problems/shortcomings with the current ocamldoc/hevea output, people who are familiar with them could look into what is required to do the necessary modifications.

The `hevea` output is _almost_ okay. There's some things, like the use of tables for centering headings in the table of contents, which really should go — if one wants such things, one does them in CSS — and there are some other infelicities like the arrows being output. The biggest issue, which is actually possibly hard in `hevea`, is generating an index/ToC sidebar, which would be easy in postprocessing.

The `ocamldoc` output is very messy. It uses tables a bunch, but worse, it has a really baroque set of CSS classes, and it's hard to set things like alternative fonts or stylings well because of this.

pmetzger commented:
BTW, let me reiterate, as strongly as @gasche fears that if we do postprocessing it will become permanent, I fear that if we don't do it, we won't get a pretty manual any time soon. The way to fix this so everyone is happy is to get people who know how to fix the ocamldoc code and who know hevea to help while we have momentum on the project.

gasche commented:
Let's take a step back out of this conversation on post-processing. There is no processing in this PR and nothing to change in either hevea or ocamldoc to get the desired output.

If/when we get a PR that does require processing, we can see how things work out. Some points that might reassure you:
- @Octachron seems quite willing to give feedback and advice and is sort of our ocamldoc expert (but then, he wasn't three years ago, and never volunteered for the job, anyone can do it with OCaml knowledge and a bit of curiosity in getting problems solved).
- I just sent an email to the Hevea author, @maranget, to ask him whether we could have a git repository for Hevea where to send prospective pull requests.

Octachron commented:
>  It uses tables a bunch, but worse, it has a really baroque set of CSS classes, and it's hard to set things like alternative fonts or stylings well because of this.

I might be responsible for this baroque set of CSS classes, which also means that I could also remove/reorganize it.

pmetzger commented:
> There is no processing in this PR and nothing to change in either hevea or ocamldoc to get the desired output.

Well, sort of. As you can learn from the thread above, @charlesetc can improve the CSS for the main parts of the manual most of the way without postprocessing. However, to fix the `libref` portions, we need to fix the `ocamldoc` generated classes. Otherwise it turns out to be quite hard to clean up the CSS.

However, in the spirit of "one step at a time", I'd be happy if we fix the main parts of the manual with just the small CSS changes in question and left fixing the rest for another PR.

bluddy commented:
Another potential question to consider is, should hevea be used at all. It's far from clear to me that the very heavy syntax of latex is the right choice for a user manual. Most systems (such as [Sphinx](http://www.sphinx-doc.org/en/master/)) use a far lighter format, such as RestructuredText, and have output in latex and html. The ocamldoc format itself should be capable of recreating everything in the manual, and that would reduce us to just one tool.

pmetzger commented:
I think a migration to Sphinx would be an excellent idea. The Coq project moved to Sphinx and it has made their lives a lot easier. _However_, I'm a big believer in doing things incrementally, and we can make the manual look much better with very little work right now. I'd suggest we do the small changes for that, and then consider arguing over a move to Sphinx later. (For the record, if an argument to the effect that we should move to Sphinx comes up, I'll likely be on the pro-Sphinx side. But I think we need to eat the elephant in small pieces.)

dra27 commented:
I completely agree with other devs on not post-processing here.

@pmetzger: - ocamldoc may be a fearfully large looking codebase, but grep is your friend. A few second's search reveals that ocamldoc/odoc_html.ml is, not too surprisingly, where all this is generated - that file consists of relatively straightforward-looking printf-like statements. IIRC the manual compiles using the ocamldoc contained in the build tree, so it looks fairly easy to me to work on changes to the HTML generator. For a more advanced solution, you could try adding a functor in ocamldoc/odoc_gen.ml to have some kind of NewHtml backend, in order to maintain compatibility, but that's not necessary to start with.

> The biggest issue, which is actually possibly hard in `hevea`, is generating an index/ToC sidebar, which would be easy in postprocessing.

Pausing slightly at the observation of "actually possibly", and with the caveat that I haven't actually looked at its code, it is very surprising to me that a program which must process a LaTeX table of contents in order to convert it HTML could have any difficulty doing this...

pmetzger commented:
> Pausing slightly at the observation of "actually possibly", and with the caveat that I haven't actually looked at its code, it is very surprising to me that a program which must process a LaTeX table of contents in order to convert it HTML could have any difficulty doing this...

I'm sure it is possible. It's just that I don't see immediately how to do it without changing code rather than configuration. That said, I may be quite wrong.

pmetzger commented:
By the way, I've put a quick-and-dirty Lambda Soup hack to postprocess the arrows into nicer SVG here: https://gist.github.com/pmetzger/4eaf8403a0e2cd3f6176d9f7367c8176

Results of postprocessing the ocamldoc.html file can be downloaded from here: https://gist.githubusercontent.com/pmetzger/bf31530b572d2a071efd459004874dee/raw/9ebab1eda3296b87ba68ae8bd92d84a24cbbe5b8/ocamldoc.html

If `hevea` can be made to generate the SVG arrows directly, then this sort of thing will not be necessary.

pmetzger commented:
Also, Steffen Smolka, in [this thread](https://discuss.ocaml.org/t/new-styles-for-code-examples-in-the-ocaml-reference-manual/1947/12), came up with a nice grey background color scheme for source code examples in the main (not libref) part of the manual. (Note that other proposed schemes are also in that thread, which was started by @Octachron, who came up with some interesting ones of his own.)

```css
@import url('https://fonts.googleapis.com/css?family=Source+Code+Pro');

div.toplevel {
  padding: 1ex 1em;
  background-color: #EEE;
  font-family: 'Source Code Pro', monospace !important;
}

div.caml-input {
  color: #11A;
}

div.caml-input::before {
  content: "#";
  color: black;
}

div.caml-output.ok {
  color: green;
}

div.caml-output.error {
  color: red;
}
```

charlesetc committed:
Select main title better; scroll overflow for tables

pmetzger commented:
By the way, enough people in discussions on IRC and discourse seem to have concerns about availability of fonts offline that it is probably going to be necessary to bundle/serve the fonts, rather than to use the Google Font CDN for it.

charlesetc committed:
Host fonts in the manual, add line to Changes

charlesetc commented:
So I've included @Octachron's suggestion to add a .maintitle class and I've also added classes to the tables that were overflowing (in the current version too), so now they have a horizontal scroll bar. I also went back to hosting fonts on the ocaml servers instead of google's, and have better font support in general.

I think the light grey for text boxes looks good on a white background, but I like the beige that's in the preview currently, since the body also has a light grey background. What do people think?

With respect to future plans, I totally agree that moving to Sphynx or another documentation generator would be great. It also might be nice to be able to release the manual changes more rapidly than ocaml compiler changes? Maybe a "current" manual or something, I'm not sure.

gasche commented on manual/manual/manual.hva:
```diff
@@ -1,4 +1,3 @@
 \input{book.hva}
-\input{fancysection.hva}
```
I should have asked before, but why was this include removed? Is it purely doing section styling?

Octachron replied:
Yes, `fancysection.hva` is only doing sections and table of contents styling.

gasche commented:
I merged this PR, and also cherry-picked in 4.07 ( 833abd351925612a2c000a927e8852df763a7fdf ).

Thanks for the nice work!

gasche commented:
>  It also might be nice to be able to release the manual changes more rapidly than ocaml compiler changes? Maybe a "current" manual or something, I'm not sure.

We have discussed this before. In principle there is an agreement that this is a good idea, but in practice we have never really felt the need to do special documentation updates. With a six-month release cycle, the manual is already updated somewhat frequently (for example, 4.07 is supposed to be released this month or shortly after). More frequent updates cannot be completely automated (we don't want an update the last-release manual to include documentation of library functions or language features that are only available in the development version).

In summary, please feel free to request an update to the manual/documentation whenever you feel this is justified by a big enough change.

dbuenzli commented:
@gasche I understand your desire to hastily merge this to avoid further bikeshedding. However I would like to suggest to @charlesetc that a 19px for the body text is way too large.

Please have a look at many websites where you read a lot, (e.g. wikipedia, github) the font size for the body text is never that large. The information density one gets on a laptop on a screenful with the the proposed font size is nearing zero. 

Why not stick to the browser default of 16px or, even better, respect the browser's user preference and use `1em` and a relative design ?

pmetzger commented:
19px is not too large I think — I sometimes argue for larger — but it is true that it's useful to rig things up to scale depending on the size of the browser. There's an excellent example of this in the CSS for [Butterick's book Practical Typography](https://practicaltypography.com/), with the point size and width being adjusted depending on the pixel sizes.

All of this is detail of course, and can be improved with time.

pmetzger commented:
BTW, is there a version of the current style up for viewing or do I have to build the manual from source to look at the latest version?

pmetzger commented:
Assuming for the moment that the version that's up is the preview, there are still a bunch of nits we probably need to fix. For example, the table that put out "type spec" in this: http://preview_docs.narwhal.space/libref/Arg.html is pretty eccentric, with the comments getting far too much width while the actual types get almost no space.

We might also want to bring in EB Garamond as a specified body font for the main part of the manual rather than leaving it to whatever the browser happens to have around as a default.

dbuenzli commented:
> All of this is detail of course, and can be improved with time.

None of this is detail. The manual became less readable for me with the merge.

gasche commented:
I merged as I believed that we were at a point where there was a global consensus that the proposed change was an improvement (in particular the use of a CDN, against which people have legitimate criticism, had been reverted). That does not imply of course that no further improvement is possible, and further PRs are welcome.

I hadn't realized @dbuenzli that the change was a regression for you, apologies. Is there a way to fix the regression that you and @charlesetc would agree on?

gasche commented:
Thinking about it more and re-reading the discussion, I indeed realize that I may have merged too quickly. Apologies. One could revert, but for now I'll leave it as is and let people submit further PRs if they wish to.

I think that it would be useful if style PRs came with a well-specified list of points that they are improving on, so that we have a well-defined scope of changes to discuss.

This PR started with a font change for the headers, and a layout change; there was general agreement that those were nice changes, reasonable concerns on the specific-font handling, and we should be able to converge quickly on something that people like. But then we started discussing background color, Lambda Soup, and the styling of code examples. This sort of feature creep not only makes discussions longer and merges slower, it makes it difficult to keep track of what the merge-ability status of the PR is.

charlesetc commented:
@dbuenzli I tend to think that larger fonts are more readable myself, but I've got 17px for the manual and 16px for the library documentation [up at the moment](http://preview_docs.narwhal.space). Does that look better?

@pmetzger the table in that Arg page doesn't look great, but it does looks better in the preview than in the current one.

@gasche that's a great point, next time I'll make a bulleted list of what the PR is trying to accomplish.

pmetzger commented:
@dbuenzli And it has become vastly _more_ readable for me and for other people, including me. No style is going to make everyone permanently happy, but we can make things better with time. For example, you could contribute some JavaScript to let you set the size permanently. It's pretty easy, you could set a cookie to set the font size and have JavaScript alter the style to suit. There are fragments online that do this we could steal quite quickly.

pmetzger commented:
@charlesetc I think the larger font was better. But I think we can do some more PRs to adjust this. Some JavaScript so that @dbuenzli and others can customize the font size to suit their tastes should be straightforward.

xavierleroy commented:
Concerning font size, what's wrong with using the browser's default (along with sizes expressed in percents or em units) and let users choose the size they want through their browser?  What looks good to you on, say, a Retina laptop is probably not what looks good to me on my vintage 30" monitor with big pixels.

pmetzger commented:
@charlesetc BTW, one other alternative here is to set the font size using `%` rather than `px`. I prefer the control px gives but `%` should work fine. We can set a size globally using `%`, and then use `rem` to control it for fonts defined afterwards, and then if someone sets a size globally in their browser this will scale off of it.

dbuenzli commented:
@charlesetc  The best practice on the web is to respect the user's browser setting (if only for people that rely on it accessiblity), and I think this is especially acute for technical documentation. Here's an excrept from [this book](http://book.webtypography.net/): 

> Use the default font size for paragraphs:
> Your starting point for paragraph text should be whatever size has been set as the default in the 
> browser. Device manufacturers will have determined an appropriate initial size, usually 16 px. Be 
> aware that your reader may have adjusted the default size to better suit their needs. You should 
> respect your reader’s wishes – if they’ve gone to the effort of adjusting the font size (especially the 
> browser default) it will be for a good reason.”

So you should start with a `font-size: 1rem` and design a relative design from there. If the web is too small for you adjust your browser's default font size. 

@pmetzger Come one there's absolutely no need for JavaScript for this. Simply do a relative and responsive design that uses the browser defaults.  You can then even use your browser zoom in/out feature (which is remembered accross visits to a site) to scale the design without problems.

pmetzger commented:
> Concerning font size, what's wrong with using the browser's default (along with sizes expressed in percents or em units) and let users choose the size they want through their browser? What looks good to you on, say, a Retina laptop is probably not what looks good to me on my vintage 30" monitor with big pixels.

So I do agree with the notion of using `rem` rather than `px` to allow users to scale the entire font set on their own.

That said: most users (and by this I mean well over 99%) will not adjust sizing or such parameters on their own under any circumstances.

It's therefore up to a designer to make sure that things look good on a wide variety of displays. There are good ways to do this, if you look at the CSS for the racket language docs, they have conditionals that basically say "if this is on a display like this, set it this way, if it's on a display like that, set it another way". We should evolve such a set of CSS rules to make sure as many people are happy as possible.

pmetzger commented:
@dbuenzli I have no problem with using `rem` and `%` instead of `px` so everything scales with the user's changes to defaults. I do have a problem with not making the default maximally readable. That said: it is absolutely a mistake to use the browser default for paragraph text sizing.

pmetzger commented:
BTW, a good online reference for design of such things is [Practical Typography](https://practicaltypography.com/), which was written by the guy who did the typography for the Racket documentation, which is generally thought of by people in the OCaml community as something to aspire to.

pmetzger commented:
BTW, just to note again: 99% of users never adjust font sizes on their displays, and are unaware that they can even do it. Heck, I have no idea how to adjust the size in Safari under iOS on my phone. I'm sure I _can_, I just don't know how, and probably never will. As a result, though it is reasonable to accommodate people who want to change sizes, it is also necessary that the daily default be good for as many people as possible.

dra27 commented:
A gentle reminder not to ignore minorities - the 1% will include people like me who have to scale things because our eyesight is rather below average.

That said, you all seem to have agreed that absolute sizes are not great...

dbuenzli commented:
> BTW, just to note again: 99% of users never adjust font sizes on their displays, and are unaware that they can even do it.

Maybe not you but people with accessiblility needs do however.

> That said: it is absolutely a mistake to use the browser default for paragraph text sizing.

I don't exactly know where you got that from. It seems that you are trying to project your own tastes as facts. The browser default is absolutely reasonable and works for millions of people on thousands of websites.

pmetzger commented:
@dra27 said:
> A gentle reminder not to ignore minorities - the 1% will include people like me who have to scale things because our eyesight is rather below average.

As is mine. (I've got bad retina problems in my right eye, and issues with focus in the left. It doesn't prevent me from driving but reading can be a chore.)

I'm _not_ against allowing people to scale their fonts. `rem` etc. seem fine. I'm against picking small ones by default.

@dbuenzli said:
> Maybe not you but people with accessiblility needs do however.

I'm someone with visual problems, so yes, I'm a person with accessibility needs. I'm in favor of using CSS unit measurements like `rem` that are based on scaling off the basic browser font values, but I'm against picking small by default. Do you now understand my position?

pmetzger commented:
BTW, just to address another one of @dbuenzli's points:
>The browser default is absolutely reasonable and works for millions of people on thousands of websites.

I just looked at a half dozen newspaper and documentation web sites, to confirm what I thought was true, and all specified font sizing (though many with `rem` and `%` rather than `px`).

The default font selection and font size in browsers is a historical accident, not something that should be relied on to give good results.

BTW, you were speaking of people with accessibility needs — those of us with poor vision are people who don't appreciate small default font sizes, as you might guess, and the defaults are generally too small.

gasche commented:
Would someone be kind enough to propose a new PR to use `rem` consistently instead of `px` settings for font sizes? We could still have a discussion about the `font-size: X%;` choice for the root element, but unfortunately I have absolutely no idea of how to resolve your fundamental disagreement? (Would the people that cared enough about the topic to comment on this PR thus far just *vote*? Do we have a Benevolent Style Dictator?)

pmetzger commented:
> Would someone be kind enough to propose a new PR to use `rem` consistently instead of px settings for font sizes?

If @charlesetc is willing I think that would be best, but if not I'll happily do that.

I'd suggest we think of the style as a work in progress. It will require a number of adjustments to make sure we account reasonably for different viewing devices, to adjust problems we find in particular pages, etc., and doing this over a series of Pull Requests will work best.

Meanwhile, @charlesetc did us all a service by showing us a much nicer looking page that seems like a good basis for future work and I greatly thank him for doing what he's done so far. Bravo.

charlesetc commented:
I'm also in favor of some sort of relative font sizing — seems everyone is. I'm not sure how `rem`'s work so @pmetzger do you think you could do the PR? Let me know if I can help!

Also thanks to all who helped review!

pmetzger commented:
@charlesetc So a `rem` is just like an `em`, except it remains constant throughout the document, while `em` changes in value to whatever the most recent `font-size` declaration stated. For most users, a `rem` is equivalent to 16 points throughout (unless you set a new root `font-size` in the style), while an `em` might shift up and down.

This explains it pretty well: https://zellwk.com/blog/rem-vs-em/

Want to give it a try?

charlesetc commented:
Thanks! Seems pretty simple and I put up a PR ^^

jasim commented:
@charlesetc This is very good stuff. Thanks a ton!

charlesetc commented:
So things look pretty good! (Also thanks @jasim :))

I noticed that the fonts are not being hosted. I would expect them to be at https://caml.inria.fr/pub/docs/manual-ocaml/fonts and that's indeed where it seems the css is looking. They still render if the computer has Fira cached or installed, but otherwise not. Maybe someone who knows the details of how the manual is hosted might be able to help?

gasche commented:
I guess that the `release` target in `manual/manual/Makefile` needs to be fixed to also copy over fonts files to be included in the manual tarball.

xavierleroy commented:
Why host a copy of the fonts on caml.inria.fr or embed them in the manual distribution when we can fetch the fonts straight from Google?  fonts.google.com recommends:

To embed your selected fonts into a webpage, copy this code into the <head> of your HTML document.
```
<link href="https://fonts.googleapis.com/css?family=Fira+Sans" rel="stylesheet">
```
Use the following CSS rules to specify these families:
```
    font-family: 'Fira Sans', sans-serif;
```

gasche commented:
@xavierleroy there are pros and cons, the main cons being that some people do not appreciate having to query a Google-owned webpage for non-Google-related errands, as they want to avoid Google web-tracking. There has been a lengthy debate on the Discuss thread related to the manual-change PR and the consensus was that hosting the fonts to make the manual self-contained is a reasonable compromise. (As far as I understand, there are no licensing issues with distributing the files ourselves.)

xavierleroy commented:
I must have missed that debate.  Some other people may not agree to serve fonts from caml.inria.fr either, you know.  That's not our job.

At any rate, the semi-absolute URLs `/pub/docs/manual-ocaml/fonts/fira-sans-v8-latin-regular.*` currently hard-coded in style.css are pessimal.  On the one hand, they force the fonts to be embedded in the tarball for the HTML manual distribution, because http://caml.inria.fr/pub/docs/manual-ocaml is populated by untarring this distribution.  On the other hand, since they are not relative URLs, they won't work for a local installation of the manual or an installation on a different Web server.

alainfrisch commented:
Re web-related tracking, @gasche perhaps referred to GDPR compliance issues ( https://github.com/google/fonts/issues/1495 ).

pmetzger commented:
> At any rate, the semi-absolute URLs /pub/docs/manual-ocaml/fonts/fira-sans-v8-latin-regular.* currently hard-coded in style.css are pessimal. On the one hand, they force the fonts to be embedded in the tarball for the HTML manual distribution, because http://caml.inria.fr/pub/docs/manual-ocaml is populated by untarring this distribution. On the other hand, since they are not relative URLs, they won't work for a local installation of the manual or an installation on a different Web server.

Yah, the fonts should be handled as relative paths etc.

gasche commented:
@pmetzger or @charlesetc: would one of you be kind enough to send a PR that uses relative paths in the right place, and fixes the `manual/manual/Makefile release` target so that the tarball contains the fonts in the right place? I can refresh the 4.07 documentation with a newly-created tarball from it.

@xavierleroy: Personally I don't see this as an issue, we serve CSS style sheets, HTML pages, and gif/svg images of navigation arrows, so why not (open source) font files. If you strongly oppose serving font files, personally I would suggest rolling back on the use of alternative fonts altogether. There is little point in binding the manual to content that only third-parties can serve.

xavierleroy commented:
Font files can be large (I haven't measured how large) and are copyrighted.  For those two reasons I'm reluctant to put them in the distribution tarball for the HTML manual.

Serving a copy of the font files from caml.inria.fr is much less of an issue.  However, local installations of the HTML manual will either miss the font files (if relative URLs are used), or fetch them from caml.inria.fr (if absolute URLs are used).  In the latter case, tracking can occur, just not from Google.

CSS have fallbacks for font declarations, so I have no problems using Fira if it's in the browser and some other font(s) as fallbacks.

pmetzger commented:
Just an aside: one of the nice things about using open source fonts is that we're allowed to use them any way we like. They can be served by google, or INRIA, or a CDN, or available locally, as we prefer. We can pick any method that suits us.

jasim commented:
This is a font-family stack that can work well across operating systems: 

    font-family: system-ui, BlinkMacSystemFont, -apple-system, Segoe UI, Roboto, Oxygen, 
                 Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;

It comes from https://tailwindcss.com/docs/fonts/#sans-serif

Happy to help with a PR if it will work.

pmetzger commented:
@jasim We have no reason whatsoever not to use a font file. Webfonts have been a thing for a while, and we get to precisely control what the user sees by selecting one. Given that the fonts are open source and that we can supply them at will, there's no incentive not to. They're not even very large. (This has also been beaten to death.)

gasche commented:
For the record, the Fira Sans files (I obtained them from [this website](https://google-webfonts-helper.herokuapp.com/fonts/fira-sans?subsets=latin)) are 192Kio in total:

```
28K	fira-sans-v8-latin-regular.eot
56K	fira-sans-v8-latin-regular.svg
56K	fira-sans-v8-latin-regular.ttf
28K	fira-sans-v8-latin-regular.woff
24K	fira-sans-v8-latin-regular.woff2
192K	total
```

In comparison, the PDF version of the manual that we serve from caml.inria.fr is 1.8Mio, and the HTML source of the ["language extension" chapter](http://caml.inria.fr/pub/docs/manual-ocaml/extn.html) is 184Kio.

gasche commented:
Report: Xavier (privately) agreed to look the other way while I install font files. We still need someone to rewrite the manual styles to use relative paths, and the `release` rule to (download and?) install the files in the right place, so that they are included in the archive we distribute. Then I can regenerate the archives and upload them on the website. I don't mind doing the path-rewriting and Makefile change myself, but in that case it will wait until the end of August.

Note: from the root of the repository, the better way to run the release script (after building the distribution with `make world.opt -j5`) is 

```
  cd manual
  RELEASENAME=4.08.0+dev make release
```

This puts the manual files in `$HOME/release`, which is an inconvenient hardcoded path, and you might have to `mkdir -p` this directory beforehand, but this is how it is right now.

dbuenzli commented:
Just want to add that you might also just go with `woff2` rather than the whole browser compat menagerie of font files, see [here](https://caniuse.com/#search=woff2) for current support. If you find that too scary (remember though that this will gracefully degrade if the browser doesn't support it) I would just go with `woff+woff2`, see support for `woff` [here](https://caniuse.com/#search=woff).

pmetzger commented:
BTW, side note: https://github.com/BlackFoundry/InriaFonts

I'm amused that the Black Foundry refers to this in their documentation as "the Inria type system".

pmetzger commented:
(Sadly, the Inria type system lacks a monospace font!)
